lib/ts/recipe/totp/recipe.ts

```diff
@@ -0,0 +1,217 @@
+                pathWithoutApiBasePath: new NormalisedURLPath(CREATE_DEVICE_API),
+            },
+            {
+                id: VERIFY_DEVICE_API,
+                disabled: this.apiImpl.verifyDevicePOST === undefined,
+                method: "post",
+                pathWithoutApiBasePath: new NormalisedURLPath(VERIFY_DEVICE_API),
+            },
+            {
+                id: VERIFY_CODE_API,
+                disabled: this.apiImpl.verifyCodePOST === undefined,
+                method: "post",
+                pathWithoutApiBasePath: new NormalisedURLPath(VERIFY_CODE_API),
+            },
+            {
+                id: REMOVE_DEVICE_API,
+                disabled: this.apiImpl.removeDevicePOST === undefined,
+                method: "post",
+                pathWithoutApiBasePath: new NormalisedURLPath(REMOVE_DEVICE_API),
+            },
+            {
+                id: LIST_DEVICE_API,
+                disabled: this.apiImpl.listDevicesGET === undefined,
+                method: "get",
+                pathWithoutApiBasePath: new NormalisedURLPath(LIST_DEVICE_API),
+            },
+        ];
+    };
+
+    handleAPIRequest = async (
+        id: string,
+        req: BaseRequest,
+        res: BaseResponse,
+        _: NormalisedURLPath,
+        __: HTTPMethod
+    ): Promise<boolean> => {
+        const options = {
+            config: this.config,
+            recipeId: this.getRecipeId(),
+            isInServerlessEnv: this.isInServerlessEnv,
+            recipeImplementation: this.recipeInterfaceImpl,
+            req,
+            res,
+            appInfo: this.getAppInfo(),
+        };
+        if (id === CREATE_DEVICE_API) {
+            return await createDeviceAPI(this.apiImpl, options);
+        } else if (id == VERIFY_DEVICE_API) {
+            return await verifyDeviceAPI(this.apiImpl, options);
+        } else if (id == VERIFY_CODE_API) {
+            return await verifyCodeAPI(this.apiImpl, options);
+        } else if (id == REMOVE_DEVICE_API) {
+            return await removeDeviceAPI(this.apiImpl, options);
+        } else if (id == LIST_DEVICE_API) {
+            return await listDevicesAPI(this.apiImpl, options);
+        }
+
+        return false;
+    };
+
+    handleError = async (err: STError, _: BaseRequest, __: BaseResponse): Promise<void> => {
+        throw err;
+    };
+
+    getAllCORSHeaders = (): string[] => {
+        return [];
+    };
+
+    isErrorFromThisRecipe = (err: any): err is STError => {
+        return STError.isErrorFromSuperTokens(err) && err.fromRecipe === TotpRecipe.RECIPE_ID;
+    };
+
+    getUserIdentifierInfoForUserId: GetUserIdentifierInfoForUserIdFunc = async (userId, userContext) => {
+        if (this.config.getUserIdentifierInfoForUserId !== undefined) {
+            const userRes = await this.config.getUserIdentifierInfoForUserId(userId, userContext);
+            if (userRes.status !== "UNKNOWN_USER_ID_ERROR") {
+                return userRes;
```

Comment:
need to add test to make sure that TOTP recipe can be used with a custom user ID that is not known to supertokens